### ISSUE TITLE
Secondary role issue

### ISSUE THREAD

**Issue by cbucknell** [UltimateHackingKeyboard/firmware#207]
- Firmware: 8.5.4
- Agent: 1.2.12
- OS: Windows 10

To replicate this issue, assign the "Mod" layer as a secondary role to the right spacebar, using Space as scancode (default). This enables the use of the arrow cluster with one hand, while still being able to use the right spacebar as Space when tapping.

Issue: Whenever space and another key are quickly pressed alternately, most spaces are not registered (e.g., alternating "a" and space: _aa a a aa a a aaaaaa_). Issue can be also replicated when typing very fast since some spaces are not registered between words. This issue is solved when the secondary role is deactivated (e.g. _a a a a a a a a a_).

Additionally, when enabling the secondary role through a third party software (TouchCursor: https://martin-stone.github.io/touchcursor/) and disabling the secondary role in the UHK, the issue disappears. This suggests that the issue is firmware related.

**Comment by mondalaci** [comment]
@kareltucek @eltang Can you guys reproduce this, or have an idea about the root cause? This hasn't ever been reported yet.

**Comment by kareltucek** [comment]
If I understand the issue correctly, it's the same old good one which has been discussed between me and @p4elkin ( https://github.com/UltimateHackingKeyboard/firmware/pull/193 and  https://github.com/UltimateHackingKeyboard/firmware/issues/187 ). 

Basically, when writing fast, the keypresses overlap in time. I.e., when writing " a", the a is pressed prior release of the spacebar which therefore acts as a layer switch instead.

So yes and yes. Unfortunately, I do not have a good idea about a proper solution... 

/E: removed the no-longer-relevant paragraph

@cbucknell, would you mind investigating and describing here how the touchcursor solves the resolution? I didn't manage to get it running in a virtual machine...

/E: sorry for the edit mess... I hope that email notifications are not fired at every update?

**Comment by cbucknell** [comment]
@kareltucek managed to find the source code of TouchCursor: https://github.com/martin-stone/touchcursor -- is it useful?

**Comment by kareltucek** [comment]
Ok, once again, seems like I've let my expectations mislead me into a terribly bad analysis.

I went through the source code and tried the program on an old windows box. The touchcursor solution does not contain any explicit delays at all - it simply resolves the situation depending on the first key up event (if the "layer switcher" is released first, it is taken for primary action, otherwise for secondary). Activation of the "action key" seems to be postponed until its release. If both keys are held without release, it seems that windows repetition hook kicks in on its own.

The touchcursor handles it via a relatively simple state machine, which is defined at 493 in https://github.com/martin-stone/touchcursor/blob/master/touchcursordll/dllmain.cpp . The postponing mechanism uses a single register. Nice & clean.

This strategy sounds very reasonable to me. The only imperfection is a small delay before activation of "held" keys (like mouse movement). 

As a result I gave one more though to the mechanics of my modified macro engine - this time I've found a way how to integrate a proper solution without much pain. Lets view it as a proof-of-concept experimental implementation. Here is a prepared configuration so you can try out easily:
[UserConfiguration.json.txt](https://github.com/UltimateHackingKeyboard/firmware/files/2822495/UserConfiguration.json.txt)
https://github.com/kareltucek/firmware/releases/tag/v8.5.4.kt.4

After flashing the firmware and loading the configuration, you shoul be in an "INI"  keymap. Then, hitting spacebar will initialize internal variables and switch you into M3A - a regular US qwerty layout. Here, S D F act as secondary switchers - S will activate mouse mode on H J K L; D will activate arrow keys on the same cluster.

@mondalaci what do you think?

**Comment by mondalaci** [comment]
@kareltucek Very sorry for the late follow-up, and thanks for delving into this! I'm not sure when will I be able to dedicate sufficient time for this issue.

@cbucknell Would you give a try to @kareltucek's firmware to see if it resolves the issue?

**Comment by cbucknell** [comment]
@mondalaci tried @kareltucek firmware and it solves the space issue. Even if I type very fast, spaces are indeed recorded every time! Did not understand though how the SDF secondary switches work though.

Edit: did not load the config correctly, see comment below

**Comment by kareltucek** [comment]
@cbucknell , now I am not sure what exactly you have tested if you write that you did not understand the SDF switches. Did you not understand how to use the bindings or how is the macro composed or how it internally works? (The config/firmware above does not contain modified space, as you must have noticed, but contains secondaries on the three keys, so that if you press `s+j`, the cursor should move down, etc..)

If you wish to modify your own maps to use the spacebar, there's a (simpler) example an some documentation in the repository at https://github.com/kareltucek/firmware - just search for `resolveSecondary`. If you need to tailor the macro for different function or explain anything, simply ask ;-) .

/E: refined the explanation in the readme a bit.

**Comment by cbucknell** [comment]
@kareltucek @mondalaci I didn't load the config correctly and had touch cursor enabled. Sorry!

Fixed it and now the SDF switches are working! I tried alternating quickly S and J to check if sometimes the cursor would get activated. It rarely does, however it seems that it happens a bit more often compared to the touch cursor. I imagine it depends on the listen time? How complex is it to implement touch cursor's implementation so we can compare?

**Comment by p4elkin** [comment]
@cbucknell fwiw, I guess months ago I had a similar issue (@kareltucek mentioned the issues like #187). I have created my own fork which is spiritually not far from what @kareltucek has done but is much less feature rich: it only re-implements the way first/second roles are handled. Been living with it for quite some time and experience almost no troubles whatsoever. Unfortunatelly, I don't have time to actively participate on this research right now, but would like to return in future.

Here's how my typical config looks like, if that's something you are aiming for (as you can see - all the layer toggles are on alphabetic keys, as well as alt and cmd) - feel free to give it a shot. 
<img width="1240" alt="default" src="https://user-images.githubusercontent.com/1240136/53511225-9342a780-3ac8-11e9-89f4-60406455b1db.png">

This fork should be compatible with agent and the normal configuration flow should be applied.
The latest release is here: https://github.com/p4elkin/firmware/releases/tag/8.5.3.187-4

cheers, HTH

**Comment by kareltucek** [comment]
> I imagine it depends on the listen time?

It should not if the time is long enough. You can increase the time to e.g., 600ms to make sure.

> How complex is it to implement touch cursor's implementation so we can compare?

This should be (more-or-less) the touch cursor version as far as I understand the code correctly. I have to admit that I am a bit confused by the waitMappedDownSpaceUp state and by the fact that the code contains no timing checks *at all*. I think that I can see one more parameter which could be possibly added to the solution and which might be present in the touch cursor solution by means of internal mechanisms of the windows hooks. I will think about it a bit more. 

Unfortunately, I cannot take the thing from touch cursor directly and put it into UHK, because the two things take entirelly different sort of input - UHK provides raw key states, while touch cursor takes preprocessed windows hooks (I.e., KeyDown, KeyUp and KeyTap actions or something similar) and hacks around them. 

> The latest release is here: https://github.com/p4elkin/firmware/releases/tag/8.5.3.187-4

@p4elkin Cool! I will give it a shot tomorrow!

**Comment by kareltucek** [comment]
My conclusion after one day of testing is that @p4elkin's firmware works like a charm. 

Regarding my solution, I agree that there still is a difference between me and touch cursor, but I am afraid that I am not capable of implementing it fully without a heavy refactor of key state handling.

**Comment by cbucknell** [comment]
@p4elkin will give it a shot! Will provide impressions soon.

**Comment by cbucknell** [comment]
After two weeks of testing @p4elkin's firmware, I can say it works really well. It would be great to see this functionality implemented in UHK's firmware by default. Maybe even having an advanced option to modify the timeouts. cc @mondalaci

**Comment by mondalaci** [comment]
@cbucknell We'll definitely consider it. Thanks for your feedback, and @p4elkin's great work!

**Comment by kareltucek** [comment]
Btw, there's a new release of my version of the feature. (Finally, I have refactored the state handling mechanism so that it still follows the official code 1:1, but decouples activation of keys from hardware and debouncing states.) Any feedback will be appreciated.

**Comment by p4elkin** [comment]
Hey @mondalaci, 

Sry to distract with this topic once again, I've a small question: I still experience intermittent chattiness of the keys with my fork. It doens't happen consistently (which makes it super hard to trace) and _*only affects the left side*_. Do you maybe have any pointers to pay attention to (like halfs state synchronisation or whatnot)? 

It must be somehow related to the fact that I continiously track the pressed keys and sometimes include them into the report in a delayed fashion (i.e. detect the pressed key, but actually push it into the report in the next update cycle), but I am a bit clueless, cause it never ever happens with the right half (which makes me think that the algo is mostly fine, but violates some underlying communication logic).

I know that helping with forks isn't the teams responsibility, just maybe if you have anything from the top of the head, thanks!

**Comment by mondalaci** [comment]
@p4elkin Can you reproduce this issue with the official UHK firmware? This might be due to a [hardware issue](https://ultimatehackingkeyboard.com/modding-and-repair/fixing-misbehaving-keys).

**Comment by p4elkin** [comment]
wow, looks like I indeed totally mis-read the symptoms and after I took care of the most problematic keys mechanically, the situation improved dramatically. 

So far no occurrence of chattiness 🎊 👍 (could save me some time spent on banging against the wall 🤕 😀 )

**Comment by kareltucek** [comment]
Sorry for being pessimistic, but if this was a switch-only issue, it should not have had occured on left half only. Of course, cleaning switches is one way to fix a misbehaving debouncer :-).

My suggestion is to look at the way the key states are rewritten into state matrix - as far as i remember, right half is synchronized synchronously ( from usb_report_updater.c 463 ), while left half is synchronized asynchronously ( similar snippet in  slave_drivers/uhk_module_driver.c ) - from an interrupt/callback. 

If the synchronisation occurs in the middle of processing the key in the updateActiveUsbReports function (I.e., between debouncing and between updating last to current.), it can mess things up.  I am no longer sure whether the problem affects official firmware too - if it does, then only with very rare occurences since the problematic timespan is very short within official firmware. Either way, even an innocent change into the code can trigger this...

**Comment by p4elkin** [comment]
my "usual suspects" were 'b' and 't' (most of the issues). I was almost sure it was a mechanical issue, but then I had several cases of other keys on the left half mis-behaivng, which made me think of the programmatic source of the problem. Anyhow, so far it seems surely bearable, but I'll keep in mind your pointers, will be the first thing to dig if the problem comes back in full power, thanks.

**Comment by p4elkin** [comment]
@kareltucek your pessimism was prophetal - the issue kept coming back randomly (sometimes - annoyingly). 

However, in my mind I had ruled out the potential "race" condition of key state matrix update: didn't know at all how interrupts work and thought that the asynchronity achieved in a similar way to e.g. JS frameworks (you have some sort of a loop which may trigger an async callback only after the user code returns control, which lead me to think that while another iteration of usb_report_updater is running - the key state matrix is intact). Looks like I was wrong and indeed what you've described appears to be a plausible scenario.

I did apply the synchronisation: whenever the communication happens over i2c (in uhk_module_driver) - update not the `KeyStates` itself, but a dummy copy (luckily memory can fit it in :)). Then in usb report updater read the left half's current states from the dummy, pretty much the same way like the right half's keys are updated. 

Never had a single occurrence of chattiness ever since (around a day of active work so far)!

Thanks again for the suggestions so far, will update if I notice the problems later.

p.s. @mondalaci seems to me that the official firmware might also be vulnerable to this problem at least in theory, if it might give you some food for thought, here's the commit: https://github.com/p4elkin/firmware/commit/771a9511f9000cc6e4a75f0240647e77aa10d32a (obviously far from an optimal solution, but does the trick just fine).

**Comment by kareltucek** [comment]
Well, what if we assume that race and bounce happen both at the same time? I think that it could actually trigger the problem in official firmware.

Some ideas about testing this:
- Set debouncing period to absurdly high value (5 seconds?) and simule bounce by actual doubletap  - the longer the release time between keys the better.
- Move debouncer into its own loop before the internal loop begins in order to increase the probability of race.

**Comment by p4elkin** [comment]
The second bullet point is exactly what I've been doing in my fork: before emitting any report updates - mitigate bouncing, populate the active state, see the situation with the pressed keys vs their previous state etc, only then - process the state and emit the actions. 

During one of my investigations of the matter I've been stripping out all the logic I had to barely this: populate a simple state and emit report records re: currently pressed keys. And I was occasionally experiencing chattiness (now makes sense kinda!).

Re: first point - nice suggestion, was also thinking to give it a shot that way. I went ahead and even though it performs a tad weirdly, there's a clear difference between how the halves perform: I never managed to screw up the right one, whereas with the left one I can often make it type several identical chars during the large debounce period.

Here's a branch in my fork that simulates the situation: https://github.com/p4elkin/firmware/commits/simulate-sync-issues, here's the commit: https://github.com/p4elkin/firmware/commit/f70cd1e18133cac79362f0a2b4694cfc0c32f561

**Comment by p4elkin** [comment]
What is also interesting, this commit: https://github.com/p4elkin/firmware/commit/573a2ef1dcbd0a9b60fa8e9531005c6a1988fd72 (the same trick that I applied to my fork) gets the situation back to normal in my case and the halves start to behave in the same way!

I'll probably go ahead and file a separate ticket for this.

**Comment by control-freak** [comment]
can anyone please tell me how to flash the firmware and load the configuration of kareltucek? i am fairly new to this and googling didn't help at all.

**Comment by kareltucek** [comment]
Sure! 

1) Note that this thread is pretty old and therefore most of the information are no longer relevant.
2) Flashing:
  - First, download new release from https://github.com/kareltucek/firmware
  - Start up your agent - any version should do. Of course, preferably a newer one (if I remember correctly, compression format was changed at some point, and I am not sure if the backward compatibility applies).
  - Click firmware in the left pane.
  - Click 'Choose firmware file and flash it' (and select the file)
3) Writing your first macro should be covered by 'Getting started' at https://github.com/kareltucek/firmware. If not, please create a new ticket in that repo and tell me there what can be improved.
4) Regarding secondary roles, I see that there is no relevant example, but something like this should do (each line has to be pasted as a separate "write text action"):
```
$ifSecondary final holdLayer mod
$tapKey escape
```

**Comment by control-freak** [comment]
So can this be be achieved with better software? Sorry i probably sound stupid but can you give me a link to download the agent? Google gives too many different results.

**Comment by kareltucek** [comment]
Agent is the configuration tool for UHK. E.g., you can get it at https://github.com/UltimateHackingKeyboard/agent/releases

**Comment by control-freak** [comment]
It says  **Cannot find your UHK
Please plug it in!**  I have a normal keyboard. Will that do?? :|

**Comment by kareltucek** [comment]
Nope. But do not despair, you can buy your own UHK at https://ultimatehackingkeyboard.com/ :-).

**Comment by eric-fulfil** [comment]
I desperately desire UHK to be the best keyboard in the world in all ways (99% there!) and the implementation of Secondary Roles has 2 major issues that make it useless for all my use-cases when compared to ZSA's firmware for example, which does this perfectly.  I wish to use 'A' and ';' as shift keys (and other modifiers under my fingertips, but we need not get that crazy yet), but typing at even a moderate speed is unusable in this case, typing "back" almost always results in "bCk".  The fix here is simple (in theory, the firmware code is a bit beyond my reasoning) and the only thing stopping me from $ on 2 more UHKs!  Top of this wishlist is changing or providing options for:

1. The timeout after a key is released before it's secondary role potential is cancelled and the primary button fires.  For all my cases, this timeout must be exactly zero; once the key is up, shift should not be applied.  I think the timeout here is 250ms or something like that, but I can't even think of a use-cases where non-zero even makes much sense tbh.
2. The primary key should not be fired after enough time has elapsed to enter secondary mode.  I might want to shift then change my mind after awhile, without wanting to type 'A'.

**Comment by kareltucek** [comment]
Eric, we are talking about the basic simple implementation which is present in the official firmware, right?

Have you tried the implementation in p4elkin's firmware or mine? If not, you definitely should. (If you have any questions or ideas about improving the algorithm provided on my branch, please create a new ticket there.)

> The fix here is simple 

The fix is simple just under the assumption that you always release the previous key before pressing the next key when writing. Do you really write that way? I personally overlap presses of two to three consecutive keys when writing fast.

>  I think the timeout here is 250ms or something like that, but I can't even think of a use-cases where non-zero even makes much sense tbh.

There is no such timeout. 

The only present timeout is the debouncing interval, which basically makes every keystroke last at least 50ms, which is however much less than a regular keystroke takes (which is 90-250ms according to my measurements), so it should not affect the behaviour in any way.

------------------------

Regarding this feature availability on the official firmware, it is basically blocked by UltimateHackingKeyboard/Agent#562, since it needs configuration. (This is because due to its nature, secondary resolution is not suitable for all usecases - namely, because it is less deterministic than the simple version and because it introduces artificial delays in keyboard operation.)

**Comment by mondalaci** [comment]
Would the artificial delay introduced by the fix of this issue be noticeable by users? If not, changing the user configuration format may not be required.

**Comment by kareltucek** [comment]
Yes, it would be. (The resolution depends on the release of the previous key, so the delays may be even up to 300-400ms long.)

I am confident that in case of my implementation, leaving the users with my implementation only (assuming that it is a sane default) is not acceptable. The user must have at least a checkbox which will read something like `enable alphanumeric-friendly resolution`.

**Comment by mondalaci** [comment]
300-400 ms is significant, indeed. I'm interested in getting this fixed, and we can extend the user configuration accordingly. Waiting for Eric's testing results.

**Comment by eric-fulfil** [comment]
Hm haven't tried flashing a custom build of the firmware, was only browsing source code, but I'll give it a try, thanks for the tips.  To reply to @kareltucek I'm 100% sure that releasing a key still keeps it in secondary mode for some time, I was absolutely sure to release the key before pressing the next key, but no dice.  And just to be clear my use-cases would require exactly a zero timeout; rapidly pressing/releasing a key would immediately type the normal char upon release and deactivate secondary mode.

**Comment by kareltucek** [comment]
> but i'll give it a try

(Quite detailed instructions are just a few posts above - the alternative version of secondary roles is available via macros...)

In case you want to play with timeouts (I suspect you will want shorter than default ones), search the README for `resolveSecondary`.

>  I was absolutely sure to release the key before pressing the next key, but no dice.

I don't think I am able to reproduce the problem. If I try to coordinate release of one key with hit of the next one, I am getting mixed results. But in that case, I usually manage just to release pressure on the key, but not to lift the finger before the next hit. And that is by definition correct.

Of course, it is theoretically possible to write some machinery which will prevent short enough overlaps... but I do not see much sense in it, since I believe most people are "lazy writers" who release keys lazily just when they need the fingers elsewhere and even not necessarily in correct order...

**Comment by eric-fulfil** [comment]
I just wrote a small javascript snippet and realized that yes, I am indeed holding down the key when pressing the next key and that there is no "timeout" from releasing the key to deactivating secondary mode, my mistake.  This pretty much just leaves an optional timeout for cancelling the primary button if held long enough as a feature request.  I guess you've all thought far ahead of me in terms of supporting these more advanced use-cases.  Here's the test I used to debug and confirm that I indeed have fat fingers in case anyone's curious, just plop it into a frontend with jQuery on it:

```
    let lastUpTime = 0;
    jQuery("input").keyup(function(e) {
        lastUpTime = new Date().getTime();
        console.log( `${lastUpTime}: keyup: ${e.key}` );
    });
    jQuery("input").keydown(function(e) {
        const downTime = new Date().getTime();
        console.log( `${downTime}: keydown: ${e.key}` );
        console.log(`ms since last keyup: ${lastUpTime - downTime}`);
    });
```

**Comment by kareltucek** [comment]
> This pretty much just leaves an optional timeout for cancelling the primary button if held long enough as a feature request.

My implementation will automatically enter secondary state after 350ms, which in most of the use-cases should be equivalent to the desired no-op...

However, if needed, the conditions can be composed, e.g.:
```
$ifSecondary final ifNotPlaytime 350 holdKey LG-
$holdKey space
```
Or the other way around, if naive non-blocking secondary role is used:
```
$holdKey LG-
$ifNotInterrupted ifNotPlaytime 350 holdKeySpace
```
Still, this both assumes the custom firmware.

> This pretty much just leaves an optional ...

Supporting alphanumeric-friendly secondaries is a reasonable request, it is just not as simple as claimed above. Omitting some details, the approach which I am using is basically to wait to see order of release of the keys - if the order is `press secondary, press next key, release the next key`, or if waiting for this sequence is longer than 350ms, then first key is handled as a secondary, otherwise as a primary. This might need prolonged press to activate secondary, but is pretty reliable to not interfere with writing.

I will try to find time soon to port the functionality to the official branch since Laci states it as desired. Thinking about it again, in this case it might make sense to have both drivers work alongside, and discriminate between them based on the primary action scancode, if Laci insists on non-configured solution. 

Still, feedback on the `$ifSecondary` approch would be very welcome, since my own testing has strong "author-bias".

**Comment by djrenren** [comment]
I recently switched from ergodox to UHK and noticed this weirdness with dual function keys.
FWIW their default is requiring both keys be held for the timeout to get the modified result, and it always did what I expected.

They have a couple different options that describe a lot of the same suggestions in this thread. I don't think we need all of them, but they're useful for discussing what behavior we want:

https://configure.ergodox-ez.com/ergodox-ez/layouts/default/latest/config/tapping

**Comment by kareltucek** [comment]
> their default is requiring both keys be held for the timeout

Mm... Does it restrict order of the presses/releases? Specifically, does the dual-function key have to go first, or can the other key go first too (as long as the "overlap" condition is satisfied)?

**Comment by djrenren** [comment]
It does restrict the order. The dual-function key has to go first. The sequence must always be: (to use their syntax)

1. `SFT_T(KC_A)` Down (on the dual role key)
2. `KC_X` Down (on another key)


After that they have a couple of resolution options. This seems super sensible because it mimics how keys like control and shift work on a regular keyboard.

For the life of me I can't tell if release order matters on QMK. I lent my QMK keyboard to a friend but I'm getting it back later today and I can report on some experiments.

I think the big difference between QMK and UHK is that the UHK doesn't require a timeout to count as a hold. 
It appears to work like so:

1. `SFT_T(KC_A)` Down (on the dual role key)
---Some amount of time less than the tap timeout---
2. `KC_X` Down (on another key)
---Immediately output `X`---

which is unforgiving to fast typists (as noted earlier in the thread).

This is similar to the `Permissive Hold` option (off by default) on QMK which would work like:
1. `SFT_T(KC_A)` Down (on the dual role key)
---Some amount of time less than the tap timeout---
2. `KC_X` Down (on another key)
---Still less than the timeout---
3. `KC_X` Up
---Still less than the timeout, Output `X`---

In the `Permissive Hold` case, the keyup ordering does matter. It allows you to skip the waiting so long as the second key is released before the first.

I think Permissive hold could still have misfires for users. Requiring the key be held past some timeout to get the modified behavior seems reasonable. It does means you'd have to buffer key presses until either the timeout is reached or the key released.

Also there's a longer article on the implementation of tap and hold in QMK from their docs. It still has some ambiguities unfortunately but it's better than the blurbs in the configurator that I linked earlier: https://beta.docs.qmk.fm/using-qmk/software-features/tap_hold

**Comment by djrenren** [comment]
I've created a little fiddle of the sort of thing I'm envisioning. [fiddle](https://codewich.com/#v1,,,nVZtT9tIEO5XIuU_DNGptlXqoPuYkPSOEy20XJEO7sMJIXDsdbzC8VreddIczX._mX2J7SQC6QIBe3den3lmdtW6ZPCNrS.WrFAwgZd.70jh2gi8uvTgJ3iJWBXeSb8HEIsE16WqeDHHhU2_1..RLJzXacqqi0JVa2MCgC1HW7NauWJS5EuWjGAmRM6iwlno92JRSAWKL5ioKYZfT09PxzAcLqTbS.oof1xEZYmepXMB4P3rYZyxqnITIK78oBWZpYpWNmMynzMFMx3hqB3p_QMaun8YGwH0XNaqibnZNfs6Aqkixcj9BiIJL_fPbO3weOjY1omldRErLgpAMQLRb.MXuBxMglw.ag9bnYlGG3jRzX1slCgiRo5Gh7A3H6qAfXUVbUppPCfMPG_sWlOj492ASGBjvRsww7KWma.jCPTGEf2hL0_9XfWgiaxB8p4iIJy1kbETkEzdGS74fgCTaTspgxfyJ8zF3B.YgBGYEQzgg84oGLelh0OMBlTGQEYLRpUoUUci3lgHnueQIbCsQAttLUxgP8qJjRPev4dj_RQ6vIJuiMbxb0p.5PNCVGx3zyizZahLPOlW2LgbH9ZxDlFJVTXbk0pzqkkXgo172Zy4HjMCeqPR2ONsXXYZ.38JSz.7tDnMU_dpBtBJd8PEs.f70x6ptuiOWjQ3IDTPDd0JTNsKDjykzWss1tRSGfJoEa0hyisWJWuYMW2KlFfMWzJgP2LGEiwYUVBFpatACFckgvTDYePha85ThXhpOcSeOlxCLeG5ECvgClY4cCKy0ASwl_M.O1rCDC2yPR2brOXIDhk6hLB7LRJQFyZsVmMf3t79_tfd1fcv8Pn679tLuPmM_XgCX29vvoeGPDxd.6b.gQN4lfGc2cUwZ8VcZTCFU2ovs3h_.nC4w8ysNkSy.jJD_PzgGOsedHPqzItoFhWJoHY3.cCKq.y1WN1cLBJWdYGhH6TA12gZybjipQJZspinPIYCXTL8RTrkWL2M0Qjo9xIR1wvs4zBKEn3GXHOpWIGGPXtCeCfgYwLtmYcuLjny.Ob2I5eyxuJWrGQRkgPxizNMD8daWokFZCJPWgySjsb.MfaCUeqg6E4l3MXHYAua7oDXY6XG3I3UDAxrbGzNuPPbwPmIHFzQvHPG50xd4BI.nq.vEt8zYl5wPHaKptRvKhqxtmKMV4y3HWopq7ZVOJBznPP4GXPWQiaxbWM4cuirk71KoM_Bub1yDFD8KBWVrw9tCkikFhBbjyPS.DDRm00Tf0I1.sDTLy96x460zWPzToNy8wSjzpR8OpNlVGCskZQTry6cRW_6hqGzISlOn2wD2agGOgPbT_RtFTPkBeJzeffnNWaM8nbk0NeicGPvVcbGLgymbjswuHw7MeoA4amJpMWMA1F0DzNdNn_LfpOAu981E6W9cqjl34UNmM31LRdY5IolJPnujq7D1PCgBM5shR2qLx80cSXMBczWIfwjaoixQizBwa4PBnv1xcNTv0oc9DMQ5i1nqSLvZwlfAk8mAxP.YKoXh7g6bW2aTLqb.DSrlUJ7JKHBGEz_oH9nQ7Mx_Q8)

It includes a little visualization of how the buffer is working. It also allows playing around with behavior when there are multiple dual-function keys in play. The top the of the TS lets you tweak the timeout and the dual-function mappings. I defaulted the timeout to an absurdly high 2 seconds so that the behavior is easier to see, but lowering it will give a better impression of the typing experience. (i recommend 200ms)

**Comment by p4elkin** [comment]
hey @djrenren,

Didn't have the time to fully digest all the listed options, but it looks like what you have in mind does resonate with my version of the firmware (linked above). It is quite dated by now, but I am still on it and have no problems using 'a' 's' and 'd' as the mouse/mod/fn keys. 

The algorithm aims to switch between several states during the typing process:
1) `handleFreeTypeState()` free typing mode (ordinary keys), all the keys are emitted already upon key-down detection.
2) `handleSecondaryRoleReleaseAwaitState` secondary role key "suspicion" mode (key-down doesn't emit, key press is cached).
In this state we await other events: 
* when other key is pressed we look whether the cached secondary key is still pressed and if not - we first emit it retrospectively and then the currently pressed key (to respect the order). If the cached secondary key is still on - we move to state 3.
3) `handleActiveSecondaryRoleState` secondary key role is on, take it in account for all the emitted events until the key is released. 

Effectively the secondary role kicks in iff the respective key is held during a certain threshold (250ms, see `SECONDARY_ROLE_ALPHABETIC_KEYS_KICK_IN_THRESHOLD`) *or* if another key was _released_ while the modifier key is still pressed (normally never causes false positives during typing, cause you'd release the key with secondary role before the following key).

See the relevant snippet:
```
 if (State.modifierCount > 0) {
        // detect if any modifier becomes active
        //
        // the previous conditions have to be met
        // or any of the modifiers may pressed long enough
        for (uint8_t i = 0; i < State.modifierCount; ++i) {
            if (modifier(i)->keyRef.state->current) {
                if (secondaryRoleTimeoutElapsed(modifier(i))) {
                    shouldTriggerSecondaryRoleActivationMode = true;
                    break;
                }
            }
        }

        for (uint8_t i = 0; i < State.actionCount && !shouldTriggerSecondaryRoleActivationMode; ++i) {
            if (!action(i)->keyRef.state->current) {
                shouldTriggerSecondaryRoleActivationMode = true;
            }
        }
    }
```

You can check out the details [here](https://github.com/p4elkin/firmware/blob/8.5.3.187-4/right/src/usb_report_updater.c)

cheers!

**Comment by djrenren** [comment]
Hi @p4elkin. Thanks for the awesome writeup! It looks like your scheme is what QMK would call permissive hold. Seems reasonable.

I'm concerned a bit about this state machine because It's not clear how it handles multiple secondary role keys that are pending. It may be totally fine but I'm struggling with the firmware code cuz this is the first time I've really peeked into it.

What happens when you have the following case:

0ms:  `z/ctrl` down (we've pressed a dual function key)
10ms: `t` down (we've pressed a single action regular key)
20ms:  `x/shift` down (we've pressed a second dual function key)
30ms: `y` down (we've pressed a single action regular key)
40ms: `z/ctrl` up (we've released the first dual function key)
50ms: `y` up (we've released a single action regular key)
60ms: `i` down
230ms: `u` down

In theory this should emit:
40ms: `z`, `t`
50ms: `shift+y`
220ms: `shift+i`
230ms: `shift+u`

To call out why this case is interesting:
- At 40ms, we see the `z/ctrl` key is released so we can replay it and any intermediate actions UP TO the next unresolved key.  That is, we can't do anything with `y` down event at 30ms because it is blocked by an unresolved secondary role key `x/shift` which was pressed at 20ms. 
- At 50ms, `y` is released so we consider it modified by all the secondary roles currently pressed, even if they're unresolved. (You could make an argument that this should emit `ctrl+shift+y` because both roles were present on the down press, but i think the state at release will cause fewer typing mistakes)
- At 220ms we've reached the timeout for `x/shift`, meaning we can emit the queued `i` press (even though it hasn't been lifted yet). Notably, this is the only emit that's occurring as a result of time rather than in response to a key press.
- At 230ms: We're past the timeout for `x/shift` so we know it really means `shift`. This means that we can emit `shift+u` immediately on press without waiting for time or the up event.

I think these 4 cases capture most of the edge cases of the space. My biggest concerns with your state machine is how it handles these overlapping secondary roles (so the 40 and 50ms emits) and how it could handle timeout-based key emits (220ms). The latter might not be worth the effort or you might not agree it's the even a good idea which is totally fair, but it's the most responsive while still preventing mistakes which is why I'm throwing it out there.

**Comment by kareltucek** [comment]
> 220ms: shift+i

This does not seem right - isn't `x/shift` already in `shift` mode because of the `y` up?

**Comment by kareltucek** [comment]
> I've created a little fiddle of the sort of thing I'm envisioning. fiddle

@djrenren what about trying out your fiddling directly on the UHK? O:-)

Here's some code which implements the strategy as you describe it https://github.com/UltimateHackingKeyboard/firmware/pull/326 . All important logic is placed in `resolveCurrentKeyRoleIfDontKnowTimeout`. If you return `SecondaryRoleState_DontKnowYet`, it will be called again in next refresh cycle. When the event (press/release) pointer is NULL, it means it didn't take place yet. If it is non-null, it contains `.time` which can be used to determine behaviour.

Personally, I am hitting false positives a lot (with 200ms and permissive mode = off).

**Comment by djrenren** [comment]
> This does not seem right - isn't x/shift already in shift mode because of the y up?

Ah I was thinking per-key as in "if an entire key down/up cycle occurs while an unresolved dual function key is held, treat that keypress as modified, but don't resolve the dual function key until the timeout", but you're suggesting that that should change the mode of the key immediately. Makes sense. I'll have to fiddle and see what works!

> @djrenren what about trying out your fiddling directly on the UHK? O:-)

For sure! Thanks for the quick primer on the code. I'll definitely start playing with it!

**Comment by djrenren** [comment]
Sorry for the slow response, I was having issues with the windows setup, so I had to get my linux environment back in order. 

@kareltucek, are the false positives you're hitting cases where the tap behavior is selected but the held behavior was intended?
Because I've noticed your implementation does something I didn't expect. At the end of the timeout, if the key has not been used as a modifier, it treats it as the tap behavior. (This gave me so many false positives that the keyboard was hardly usable). This means you have to be quick lest you get extra input. I was thinking that if you held the key past the timeout, it would behave like the secondary role. 

The following diff fixed all my false positives:
![image](https://user-images.githubusercontent.com/634666/104963004-7cdcb800-598e-11eb-936f-cc2182f6f92c.png)

This has the downside of making it impossible to hold the key down for repeated entry. QMK "solves" this by allowing a tap then hold to behave as a held down key.

**Comment by kareltucek** [comment]
> are the false positives you're hitting cases where the tap behavior is selected but the held behavior was intended?

Nope - the other way around. I.e., secondary role (hold behaviour) often activates when writing. I am a relatively decent touch typist (~55wpm), but I am also a "lazy writer", meaning that I tend to release keys quite slowly, so my problem is not much of a surprise. For instance, in the word "star", I tend to first press the first three letters before I even start releasing first.

Still I am happy to hear that it indeed works fine for you. Means we are on the right track.

> At the end of the timeout, if the key has not been used as a modifier, it treats it as the tap behavior.

Yes, this is always the question, since the "correct" timeout behaviour depends on the key in question.

But as you figured out, it is trivial to fix/"fix" :-). 

> This has the downside of making it impossible to hold the key down for repeated entry. QMK "solves" this by allowing a tap then hold to behave as a held down key.

This should be quite easy to add too - suffices to remember the previous resolution key and time and compare it against a timeout.

----------------------------------------------
Regarding my problems, it will be quite easy to fix by lenghtening the timeout and adding one more "safeguard" timeout at the end (which relies on the fact that am quite reliable at releasing keys in almost the right order), just will have to find some time to implement it. In the end, the same algorithm should work for both, just will have one more parameter...

**Comment by djrenren** [comment]
> For instance, in the word "star", I tend to first press the first three letters before I even start releasing first.

This is fascinating to me! So cool that people have such different typing styles. And that's awesome that you have a solution that'll work for both!

**Comment by kareltucek** [comment]
> This is fascinating to me! 

Yep! Exactly why I am so relieved that it turned out that the "blind" implementation of "your" strategy actually works for you! 

------------------------------

Also, I have been thinking about a related matter on which I would appreciate some opinion.

Rationale is as follows: alphanumeric dual roles are inherently complicated to use and implement, since they require either prolonged press or strict press/release order (or both). And due to the same reason, their benefit is somewhat limited. So the idea is basically - wouldn't it be more profitable to turn the problem inside out, and stick to using standard single role keys but relax their ordering constraints?

I.e., the idea is to add a small delay (50ms?) before key activation, and during this delay, to sort queued actions by the action type - layer switchers first, macros second, modifiers third and scancodes last. As a result, the user can press (almost) all shortcuts/combos/you-know-what-I-mean as a single chord, without paying attention to correct order.

Now my biggest question is - is such idea even relevant to yours or other people's typing styles?

**Comment by djrenren** [comment]
Hmmm so i don't think it scratches the same itch. For example, with the firmware I've built locally, I have `s` set as a mod layer secondary action so I can just hold it down and use my right hand for arrow keys. Something like that wouldn't work too well with this model because it explicitly takes advantage of secondary roles. All this to say, that secondary roles are still important to my workflow.

As for this like, chording style... hmm it seems like you could have issues with fast typists. I probably wouldn't use this (not that I'm too fast a typer, I just don't really need out of order pressing). In the case of basic modifiers, we've all been trained to have them pressed first so I'm pretty comfortable with that. This reminds me of a stenographer-style approach to typing, which would be cool if I was using chords all the time, but I usually have no more than a single modifier held down at any given time (aside from Ctrl+Alt+delete).

**Comment by kareltucek** [comment]
So (If I may have one more question) - are secondary roles important for you because they save travel distance of your fingers, or because they provide you more layer switcher keys?

**Comment by djrenren** [comment]
Happy to answer questions 😃 

yeah it's all about finger travel for me. I have small hands and last year got some nasty RSI symptoms so I'm always looking for ways to reduce the work my fingers are doing. I could go all the way and use lots of layers to reduce the travel even further, but I've found it's mostly reaching for the bottom row modifier keys that puts stress on my hands. With a little creative mapping with secondary roles, I can fit most of my modifiers into my resting hand position and that's been super nice. Also UHK doesn't support enough layers for more layer switchers to be that helpful. (though it looks like it's in the works?)

**Comment by kareltucek** [comment]
> found it's mostly reaching for the bottom row modifier keys that puts stress on my hands

Makes sense.

I've managed to fit my keys into layers which I control by thumb (except for pinky's mouse/mirror). I am not using my (non-thumb) fingers at all on the bottom row. Also, I have been experimentng with 40%-ish keymaps, which means that my numbers and symbols are now all in layers. That's where it "might" make sense to start optimizing on chord convenience. (The idea is to save a fraction of a second and to think about such shortcuts as of single motions rather than two-phase motions.)

> (though it looks like it's in the works?)

As far as I know, it is planned, but hasn't been moving forward for quite a long time. At the moment the only way is to either use the solution available on my fork (macro based), or to design a workflow which will use keymap switching (e.g., just like vim's modes do).

**Comment by kareltucek** [comment]
Very sorry for taking so long. 

I consider now the PR prepared. It contains a superset of the features discussed in this thread. (Includes doubletap activation, activation by release, choice of timeout action (primary vs secondary).)

On top of QMK standard features, there is what I call "safety margin". If set to nonzero value, it virtually offsets release of the action key so that activation by release prefers primary action when both keys are released (almost) at the same time. (This makes alphanumeric secondaries compatible with typing styles similar to mine.)

**Comment by djrenren** [comment]
That's fantastic! Thanks @kareltucek for your awesome work! and the wait's been no trouble. I've been running the custom build you set up before and that's worked well enough. Excited to see it done though! :smiley:

**Comment by kareltucek** [comment]
Allright @mondalaci as requested, here are two testing tarballs. I have found one bug since, so that has been fixed too. One variant (A) is configured towards timeout-induced activation (aka press both keys and wait till something happens), second (B) towards release-induced activation (aka press modifier and briefly tap the action key while making sure not to release the modifier sooner than the action key):

[uhk-fw-8.10.8-variantA.tar.gz](https://github.com/UltimateHackingKeyboard/firmware/files/6313796/uhk-fw-8.10.8-variantA.tar.gz)
[uhk-fw-8.10.8-variantB.tar.gz](https://github.com/UltimateHackingKeyboard/firmware/files/6313797/uhk-fw-8.10.8-variantB.tar.gz)

**Comment by mondalaci** [comment]
@kareltucek Thank you! Would you please describe the use cases for variant A and B, so that people can better understand and test them accordingly?

**Comment by kareltucek** [comment]
Usecases? The usecases are the same for both - when you want to use secondary role on an alphanumeric key. The difference
is compatibility with different typing styles - basically if A doesn't work for you, try B.

There are some other configuration options (e.g., activation of long primary hold can be done (depending on config) either by doubletap or by long hold), but these should be quite sane and should cover all usecases equally well.

The usecase difference which matters is between simple and timeout strategy:

- simple strategy - i.e., the standard strategy used in the latest firmware releases - depends neither on length of the hold, nor on release order and can be activated rapidly. I.e., is reliable as long as the user can dedicate a key whose primary action will not be used in rapid sequences with overlaps.

- timeout strategy - is useful on alphanumeric keys. It is designed to suppress unintended secondary role activations on rapid keystroke sequences (e.g., regular writing). The price is that it is less reliable and less deterministic. This strategy is packed in the above two tarballs as the only available one. The difference between the two is calibration of timeouts for two different typing styles:
  - variant (A) is configured towards timeout-induced activation - i.e., to activate secondary role combo, press both keys and wait till it activates.
  - variant (B) is configured towards release-induced activation - i.e., to activate secondary role combo, press the dual role key and briefly tap the action key while making sure not to release the modifier sooner than the action key. 
    
    This one should be much more reliable in suppressing unwanted secondary activations, but some people report that they don't like it.
  
  In both cases, to activate long hold of primary action of the dual role key, doubletap the key (i.e., tap and on next press hold).

**Comment by mondalaci** [comment]
@kareltucek Thank you very much for the thorough descriptions!

Anyone interested, please try the A and B variant firmwares above, and let us know your findings. I'll merge the related PR after hearing your opinions.

**Comment by mhantsch** [comment]
>   * variant (B) is configured towards release-induced activation - i.e., to activate secondary role combo, press the dual role key and briefly tap the action key while making sure not to release the modifier sooner than the action key.

Very cool, @kareltucek , thanks a lot! I was hoping UHK would one day support this -- using alpha keys as modifiers (secondary role). I need to find time to test variantB but that may take a bit (I am pretty swamped with work and other activities at the moment).

**Comment by djrenren** [comment]
@kareltucek I've noticed an odd behavior in variant B. To illustrate I'll use my keymap where the s key is a dual function key   whose secondary role applies the Mod layer (which for this example is factory standard).

### Scenario 1 (no surprises)
*Input:* (as quick as possible)
`s` down
`i` down
`i` up

*Output:*
`Up arrow`

### Scenario 2 (surprise)
*Input:* (as quick as possible)
`s` down
`s` up
`s` down
`i` down
`i` up


*Output:* 
`s`, `s`, `i`


*Expected output*
`s`, `up arrow`


It seems like somehow repeatedly tapping the dual function key before holding it down, can result in a state where its dual function role is no longer honored. Specifically, once I reach the end of scenario 2, I can keep `s` down as long as I want and the keyboard will act as if its still on the Base layer. Is this intended behavior?

**Comment by kareltucek** [comment]
> Is this intended behavior?

Definitely not. Will look into it.

**Comment by robertofarrell** [comment]
I just wanted to say thanks for looking into this issue.  Its one of the usability issues that really affects me.  I have installed the latest firmware from kareltucek and it appears to do a good job of resolving the issue but it's still not perfect.  If there is anything I can do to help test a solution I'd be interested in helping.

**Comment by kareltucek** [comment]
> Is this intended behavior?

Or actually, on second thought, both A and B have the "doubletap to hold primary" feature turned on, so it indeed looks like the intended behaviour. 

Therefore I offer variant C - [uhk-fw-8.10.8-variantC.tar.gz](https://github.com/UltimateHackingKeyboard/firmware/files/6356692/uhk-fw-8.10.8-variantC.tar.gz)


| | A | B | C | 
| :---- | :---: |:---: |:---: |
| doubletap timeout | 200 | 200 | |
| trigger timeout | 200 | 350 | 350 |
| safety margin | 0 | 50 | 50 |
| trigger by release | false | true | true |
| doubletap to hold primary | true | true | false | 
| timeout action | secondary | secondary | secondary |

-----

> ...the issue but it's still not perfect. If there is anything I can do to help test a solution I'd be interested in helping.

Nothing easier than testing the above 3 variants and formulating what exactly is not perfect :-).

Alternatively, if you are fine with compiling the firmware, it would be great if you could also try fiddling with the values directly in the code - all constants are in `secondary_role_driver.c`...

**Comment by juanperi** [comment]
I've been using variants A and B, and both of them generate too many problems for me.
I have configured letter A with dual role with COMMAND as the secondary role.

With variant A, I think the main problem is the unwanted trigger of the secondary role. In sessions where i'm just writing, the CMD would trigger, really messing with what i'm doing (eg writing ARtificial would refresh the page as it's a CMD+R)
I don't remember exactly what my problem was with variant B. I'll install it again and report back after some time using it

**Comment by p4elkin** [comment]
@juanperi maybe if you have some time could you also give a shot at my fork maybe (see [this comment](https://github.com/UltimateHackingKeyboard/firmware/issues/207#issuecomment-467961874)? I'd appreciate any feedback as well. 

Unfortunately, I don't have the time to follow the development of this thread lately, but I'm really not having any issues with my impl of the feature (could be just cause I type too slow :D).

**Comment by juanperi** [comment]
> I don't remember exactly what my problem was with variant B. I'll install it again and report back after some time using it
The issue with B is that the secondary role fails to register sometimes.
I have key CapsLock configured as ESC on press, CTRL on hold
![image](https://user-images.githubusercontent.com/203349/116510877-e6416080-a8c5-11eb-83f5-8d8ab1a1e026.png)
If I do CapsLock+P (effectively CTRL+P), many times only registers as P. I would assume it in fact registers as ESC and then P, but not totally sure.
Same happens with CapsLock+Space (effectively CTRL+Space, my tmux leader) it fails to register as well... 
Both of these situations are avoided if I force myself to introduce a delay in the way I type.

**Comment by kareltucek** [comment]
> Both of these situations are avoided if I force myself to introduce a delay in the way I type.

Yes, the B variant is configured so that it does not interfere with writing even for very lousy writing styles. As a result, the secondary role combos need a distinct sequence. (Just hammering both keys and then releasing as with regular keyboard just does not (and I believe cannot) work.) I guess some timeout configuration inbetween might work better for you - but in that case you might be getting both types of problems at the same time :-/.

Regarding the A variant, you are obviously not lifting your fingers from keys very eagerly when writing, which is the same problem for me.

> @juanperi maybe if you have some time could you also give a shot at my fork maybe (see this comment? I'd appreciate any feedback as well.

Still, would you find some time to elaborate on how your resolution strategy works? When I tried it (eternity ago), my impression was very good, yet I was not able to read easily the logic behind...

**Comment by juanperi** [comment]
> @juanperi maybe if you have some time could you also give a shot at my fork maybe (see [this comment](https://github.com/UltimateHackingKeyboard/firmware/issues/207#issuecomment-467961874)? I'd appreciate any feedback as well.

Hey, I'm testing your fork, and I'm hitting the same corner cases that @kareltucek 's version B

@kareltucek I really appreciate your work. I think Version B is the one works best for me, even though I will have to slow myself down for some combinations that it seems i do too fast. Nevertheless, the ability to  have the secondary action this way is super valuable!!! thank you for it

**Comment by mondalaci** [comment]
Everyone who's testing this, please make sure to test @kareltucek's A, B, C firmware versions and also @p4elkin's firmware. I'd like to hear your feedback on how they all perform.

**Comment by juanperi** [comment]
One difference I found between @p4elkin and @kareltucek firmwares
 
I have a Mod layer mapping for the letter D which is CMD+Tab. As long as I keep pressing the MOD layer switcher, I can keep circling through my open windows with @kareltucek 's firmware by releasing the D key and pressing it again, but I cannot with @p4elkin 

Not sure if it's intended or not, but it looks like Karel's firmware (as well as stock firmware) maintain the modifier pressed state as long as I do not release the Mod layer key

Here you can find the definition
![image](https://user-images.githubusercontent.com/203349/116657357-8e226100-a98e-11eb-9d5a-785e4e550c2e.png)

**Comment by kareltucek** [comment]
That feature is called sticky modifiers and is present (yet undocumented) in official firmware, but probably not present or enabled in p4elkins. Nothing to do with the secondary roles actually.

(At the time when the firmwares were forked, it used to be enabled globally on all non-base layer shortcuts, which brought a lot of problems for non-alt-tab-style shortcuts, so we probably both disabled the features on our branches.)

**Comment by rafaelpirolla** [comment]
I had less trouble with variant C of kareltucek. p4elkins version seems more responsive, though.
With karel's I started having same issues with backspace (which I use as MOD+M) as it computes like a combination of enter and then the m key - since I have MOD as secondary role in the enter key.
And there is no way to produce this kind of string:
lskjflsfjdslkfjlksdfjlksdjfsdlkfjls

I would rather have accidental alphas rather than accidental secondary hits in a home row mod:
https://precondition.github.io/home-row-mods#finding-the-sweet-spot

**Comment by cbucknell** [comment]
Thanks for the different versions!

I have been using firmware version C for about a month now. Before I was
using p4elkin's version (and before, the original UHK firmware).

Compared to the original UHK firmware, version C is far better. It is in
fact very similar to p4lkin's version.

The only issue I have in both firmwares (C and p4elkin's) is the following:

I have the space bar binded to scancode *Space *and Secondary role *Mod*.
Sometimes, when I am typing, the secondary role gets activated, outputting
whatever is in the Mod layer for a specific key (instead of what's in the
base layer).

Any recommendation on what parameter modification may solve this issue?

Thanks

**Comment by kareltucek** [comment]
> Any recommendation on what parameter modification may solve this issue?

The totally honest recommendation from me is to redesign your keymap so that you don't need dual role keys.

Apart from that - It is hard to figure out what the problem is, because this depends on your typing style a lot. My guess is that one of the following applies:
- You tend to hold space for too long (more than 400ms). In that case, simply increasing the trigger timeout may help. Other variant would be to set timeout action to primary.
- Or you are releasing your space in time, but are releasing your alphas even faster. (I.e., you press space, press alpha, release alpha, and wait for more than 50ms.) In that case, either increasing safety margin might help, or disabling activation by release entirelly. 

Either way, this is going to make secondary role activation even more sluggish.

Other recommendation is to look directly into the PR (https://github.com/UltimateHackingKeyboard/firmware/pull/326/files) and read the comments in the code in order to understand what the parameters mean, and then to start debugging your scenarios using this knowledge. (If anything is unclear in the PR, do not hesitate to request comments - it is meant to be easily understandable.)

**Comment by proostas** [comment]
Hello. I have come across this issue when I tried to use 'z' and '/' with Shift as a secondary role. It is impossible to rolling type 'z' following any other letter. Quickly typing _zip zip zip_ results in _Ip Ip Ip_.

In QMK firmware they fix this issue with **Ignore Mod Tap Interrupt** option and it fixes this problem completely.

On the other hand, the following macro seems to work very well for me:

>  resolveSecondary 200 1 3
> write z
> break
> holdKey leftShift

**Comment by mondalaci** [comment]
@proostas Thanks for your feedback!

@kareltucek Would it be feasible to implement a resolution strategy that works like @proostas's macro but can be enabled for all keys with a smart macro variable?

**Comment by kareltucek** [comment]
You mean hook the native secondary role onto the `resolveSecondary` resolution strategy? Yes, that's what the PR was about. It will be available when I or someone gets to porting that functionality again. (I mean port from the abandoned branch in my repo, not from qmk.)

**Comment by mondalaci** [comment]
@kareltucek I don't know what's the difference between the mentioned `resolveSecondary` resolution strategy and making all keys behave like the macro of @proostas. I assumed the latter might be easier to implement and achieve similar behavior.

**Comment by Galathius** [comment]
Seems I've found steps to reproduce for character doubling issue, not sure if it was already mentioned anywhere above.
Just saw that this issue depends on typing style, probably it would be helpful (cc: @kareltucek)

Here is the script that I use:
```
resolveSecondary 350 50 primary secondary
primary: 
    tapKey space
    break
secondary:
    final holdLayer mod
    break
```
I have mapped this script to the right `Mod` button. Every time when I quickly click on any alphabet and then hold the `Mod` button – the last typed symbol is duplicated. 

Sometimes it's duplicated quite a lot of times, and I found that it depends on the first arg of `resolveSecondary` command. The larger this value, the more duplicates you will have.  

If I would change `350` to something like `1000` there is what I have:
1) Press and release `o` key
2) Press and hold `Mod` key
3) I get: `ooooooooooooooooooooooooooooooooooooooooo`

With a `2500` value it becomes `ooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo`.

Is there any way to interrupt the repeating of the previously typed key? Is there any magic command that I can add to my script to make it work better?


**UPD**
The issue was solved by decreasing `Key repeat rate` and `Delay until repeat` macbook keyboard configuration. I had it on the extremely "fast" level.

**Comment by kareltucek** [comment]
Yep, increasing the OS repeat delay is the recommended solution.

Theoretically, the algorithm also can be improved to not prolong non-modifier scancodes... but it is not a priority.

**Comment by mhantsch** [comment]
There's a typo in the user-guide. Under the chapter **Native secondary role - simple strategy** it says it can be explicitly set by:

`set secondaryRole.defaultStrategy advanced`

But I believe it needs to be:

`set secondaryRole.defaultStrategy simple`

**Comment by kareltucek** [comment]
@mhantsch fixed! Thanks!
